### PR TITLE
Added support for configuring Java exec path for instances.

### PR DESCRIPTION
**Pull request by craigatgoogle**

### CODE REVIEW

**Comment by craigatgoogle** [comment]
Would like to add an IT with a non-standard java exec path, but will need to figure out perms for gce-jenkins project first.

**Comment by craigatgoogle** [comment]
Addresses: https://github.com/jenkinsci/google-compute-engine-plugin/issues/91

File: src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineLinuxLauncher.java [hunk 1]
```diff
@@ -127,12 +127,14 @@ protected void launch(ComputeEngineComputer computer, TaskListener listener, Ins
...
-        logWarning(computer, listener, "Java is not installed.");
+      String javaExecPath = node.getJavaExecPathOrDefault();
+      if (!testCommand(
+          computer, conn, String.format("%s -fullversion", javaExecPath), logger, listener)) {
+        logWarning(computer, listener, String.format("%s is not installed.", javaExecPath));
       }
 
       // TODO: allow jvmopt configuration
-      String launchString = "java -jar " + jenkinsDir + "/agent.jar";
+      String launchString = String.format("%s -jar ", javaExecPath) + jenkinsDir + "/agent.jar";
```
Comment (stephenashank) [inline src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineLinuxLauncher.java]: There's a lot of duplication here between these launch methods, I think it might be good to pull out one or two helper methods into ComputeEngineComputerLauncher for checking the java install and generating a launch string. 
To tackle the issue of pathing differences, I would recommend having a base class method to retrieve the agent path that gets overridden in these subclasses, one returning `/agent.jar` and the other `\\agent.jar`, or something along those lines.

File: src/test/java/com/google/jenkins/plugins/computeengine/ConfigAsCodeTest.java [hunk 2]
```diff
@@ -51,6 +51,7 @@ public void shouldCreateCloudInstanceFromCode() throws Exception {
     assertEquals("Wrong configurations numExecutors", "1", configuration.getNumExecutorsStr());
     assertEquals("Wrong configurations runAsUser", "jenkins", configuration.getRunAsUser());
     assertEquals("Wrong configurations remoteFs", "agent", configuration.getRemoteFs());
+    assertEquals("Wrong configurations javaExecPath", "java", configuration.getJavaExecPath());
```
Comment (stephenashank) [inline src/test/java/com/google/jenkins/plugins/computeengine/ConfigAsCodeTest.java]: I think it would be good to also add an integration test which makes use of a non-standard Java exec path. It could mean being able to change the startup script which might not be trivial, but an easier thing to do might just be to use the full path of the java that already gets installed instead of the default "java" and making sure that works as well.

Comment (rachely3n) [reply src/test/java/com/google/jenkins/plugins/computeengine/ConfigAsCodeTest.java]: Startup script could work since currently the linux IT's requires the startup script to install Java.

Comment (craigatgoogle) [reply src/test/java/com/google/jenkins/plugins/computeengine/ConfigAsCodeTest.java]: Done. Thanks @rachely3n for the suggestion, that really helped :)

File: src/test/java/com/google/jenkins/plugins/computeengine/integration/ComputeEngineCloudNonStandardJavaIT.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.junit.rules.Timeout;
+import org.jvnet.hudson.test.JenkinsRule;
+
+/**
+ * Integration test suite for {@link ComputeEngineCloud}. Verifies that instances can be created
+ * with multiple Jenkins labels and that these labels are properly provisioned.
```
Comment (rachely3n) [inline src/test/java/com/google/jenkins/plugins/computeengine/integration/ComputeEngineCloudNonStandardJavaIT.java]: Description updated to test non-standard java paths?

File: src/test/java/com/google/jenkins/plugins/computeengine/integration/ConfigAsCodeTestIT.java [hunk 4]
```diff
@@ -68,14 +67,39 @@ public void testWorkerCreated() throws Exception {
...
     String name = planned.iterator().next().displayName;
 
     // Wait for the node creation to finish
     planned.iterator().next().future.get();
-
     Instance instance = client.getInstance(PROJECT_ID, ZONE, name);
+    assertNotNull(instance);
+  }
+
+  @Test
```
Comment (rachely3n) [inline src/test/java/com/google/jenkins/plugins/computeengine/integration/ConfigAsCodeTestIT.java]: Not necessarily related to this PR, but how do we wish to structure IT's for config as code. With each new feature we implement, will we always have to write an integration test for its configuration as code counterpart and should all these tests live in one file  (ConfigAsCodeTestIT).

File: src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineInstance.java [hunk 5]
```diff
@@ -30,10 +31,11 @@
 import java.util.Optional;
 import java.util.logging.Level;
 import java.util.logging.Logger;
+import javax.annotation.Nullable;
 import jenkins.model.Jenkins;
 
 public class ComputeEngineInstance extends AbstractCloudSlave {
-  private static final long serialVersionUID = 1;
+  private static final long serialVersionUID = 2L;
```
Comment (ingwarsw) [inline src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineInstance.java]: With that change all old configs will work?
I must say Im not sure of that..

If we want to keep it we should create test for migrating..

File: src/main/java/com/google/jenkins/plugins/computeengine/InstanceConfiguration.java [hunk 6]
```diff
@@ -169,6 +170,16 @@ public InstanceConfiguration(
     readResolve();
   }
 
+  @DataBoundSetter
```
Comment (ingwarsw) [inline src/main/java/com/google/jenkins/plugins/computeengine/InstanceConfiguration.java]: Annotation should not go after javadoc?

File: src/test/java/com/google/jenkins/plugins/computeengine/integration/ComputeEngineCloudNonStandardJavaIT.java [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+public class ComputeEngineCloudNonStandardJavaIT {
+  private static final String NON_STANDARD_JAVA_STARTUP_SCRIPT =
+      "#!/bin/bash\n"
+          + "sudo su -\n"
+          + "/etc/init.d/ssh stop\n"
+          + "echo \"deb http://http.debian.net/debian stretch-backports main\" >> /etc/apt/sources.list\n"
+          + "apt-get -y update\n"
+          + "apt-get -y install -t stretch-backports openjdk-8-jdk\n"
+          + "update-java-alternatives -s java-1.8.0-openjdk-amd64\n"
+          + "ln -s /usr/bin/java /usr/bin/non-standard-java\n"
```
Comment (ingwarsw) [inline src/test/java/com/google/jenkins/plugins/computeengine/integration/ComputeEngineCloudNonStandardJavaIT.java]: Would be nice to be sure that java is not in path and if we will break something in future it will not work with `java` command.

Instead of link can we move it?

File: src/test/java/com/google/jenkins/plugins/computeengine/integration/ConfigAsCodeTestIT.java (hunk 4 above)
Comment (ingwarsw) [reply src/test/java/com/google/jenkins/plugins/computeengine/integration/ConfigAsCodeTestIT.java]: Im not sure if we needs JCasC integration test of each new functionality..
If we have IC test for that feature normal JCasC test should be enough..

File: src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineLinuxLauncher.java (hunk 1 above)
Comment (stephenashank) [reply src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineLinuxLauncher.java]: Understood, the TODOs you added are enough for this PR.

File: src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineInstance.java (hunk 5 above)
Comment (craigatgoogle) [reply src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineInstance.java]: The AbstractCloudSlave instances are meant to be ephemeral, therefore they shouldn't be getting seriliazed to the config.xml unless I'm misunderstanding something.

File: src/test/java/com/google/jenkins/plugins/computeengine/integration/ComputeEngineCloudNonStandardJavaIT.java (hunk 3 above)
Comment (craigatgoogle) [reply src/test/java/com/google/jenkins/plugins/computeengine/integration/ComputeEngineCloudNonStandardJavaIT.java]: Good catch. Thanks.

File: src/test/java/com/google/jenkins/plugins/computeengine/integration/ConfigAsCodeTestIT.java (hunk 4 above)
Comment (craigatgoogle) [reply src/test/java/com/google/jenkins/plugins/computeengine/integration/ConfigAsCodeTestIT.java]: I think in this case sense it directly affects worker creation having test parity for JCasC is merited so as to catch any edge cases which may be introduced through the JCasC plugin. However, I agree in general that not all features need 100% test parity for JCasC.

File: src/test/java/com/google/jenkins/plugins/computeengine/integration/ComputeEngineCloudNonStandardJavaIT.java (hunk 7 above)
Comment (craigatgoogle) [reply src/test/java/com/google/jenkins/plugins/computeengine/integration/ComputeEngineCloudNonStandardJavaIT.java]: Done.

File: src/main/java/com/google/jenkins/plugins/computeengine/InstanceConfiguration.java (hunk 6 above)
Comment (craigatgoogle) [reply src/main/java/com/google/jenkins/plugins/computeengine/InstanceConfiguration.java]: Fixed.

File: src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineLinuxLauncher.java [hunk 8]
```diff
@@ -127,12 +128,14 @@ protected void launch(ComputeEngineComputer computer, TaskListener listener, Ins
       scp.put(Jenkins.get().getJnlpJars("agent.jar").readFully(), "agent.jar", jenkinsDir);
 
       // Confirm Java is installed
-      if (!testCommand(computer, conn, "java -fullversion", logger, listener)) {
-        logWarning(computer, listener, "Java is not installed.");
+      String javaExecPath = node.getJavaExecPathOrDefault();
+      if (!testCommand(
+          computer, conn, String.format("%s -fullversion", javaExecPath), logger, listener)) {
+        logWarning(computer, listener, String.format("%s is not installed.", javaExecPath));
```
Comment (stephenashank) [inline src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineLinuxLauncher.java]: Technically we're always looking for whether Java is installed, I think this would be better if changed to `Java is not installed at %s.`

File: src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineWindowsLauncher.java [hunk 9]
```diff
@@ -126,12 +127,14 @@ protected void launch(ComputeEngineComputer computer, TaskListener listener, Ins
       scp.put(Jenkins.get().getJnlpJars("agent.jar").readFully(), "agent.jar", jenkinsDir);
 
       // Confirm Java is installed
-      if (!testCommand(computer, conn, "java -fullversion", logger, listener)) {
-        logWarning(computer, listener, "Java is not installed.");
+      String javaExecPath = node.getJavaExecPathOrDefault();
+      if (!testCommand(
+          computer, conn, String.format("%s -fullversion", javaExecPath), logger, listener)) {
+        logWarning(computer, listener, String.format("%s is not installed.", javaExecPath));
```
Comment (stephenashank) [inline src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineWindowsLauncher.java]: See comment about wording above

File: src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineLinuxLauncher.java (hunk 8 above)
Comment (rachely3n) [reply src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineLinuxLauncher.java]: Nice catch.

Comment (craigatgoogle) [reply src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineLinuxLauncher.java]: Done.

File: src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineWindowsLauncher.java (hunk 9 above)
Comment (craigatgoogle) [reply src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineWindowsLauncher.java]: Done.

File: src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineInstance.java (hunk 5 above)
Comment (ingwarsw) [reply src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineInstance.java]: They are not ephemeral..

Try adding few slaves and restarting instance.. :)

Comment (craigatgoogle) [reply src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineInstance.java]: Ah ok, good to know. In that case I've leave the UID unchanged. Older versions of the serialized object will be instantiated with a null value which we account for in the getJavaExecPathOrDefault() method.
